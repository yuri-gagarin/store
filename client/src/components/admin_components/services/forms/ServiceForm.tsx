import React, { useState, useEffect, useRef }  from "react";
import { Button, Form, TextArea } from "semantic-ui-react";

export type FormState = {
  name: string;
  description: string;
  price: string;
}

interface Props {
  name: string;
  description: string;
  price: string;
  newForm: boolean;
  handleCreateService(data: FormState): void;
  handleUpdateService(data: FormState): void;
}


const ServiceForm: React.FC<Props> = ({ name, description, price, newForm, handleCreateService, handleUpdateService }): JSX.Element => {
  // local form state and ref //
  const [ formState, setFormState ] = useState<FormState>({ name, description, price });
  const serviceFormRef = useRef<HTMLDivElement>(document.createElement("div"));
  // input change listeners //
  const handleNameChange = (e: React.ChangeEvent<HTMLInputElement>):void => {
    setFormState({
      ...formState,
      name: e.target.value
    });
  };
  const handleDescriptionChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setFormState({
      ...formState,
      description: e.target.value
    });
  };
  const handlePriceChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setFormState({
      ...formState,
      price: e.target.value
    });
  };
  // submit listener //
  const handleSubmit = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if (newForm) {
      handleCreateService(formState);
    } else {
      handleUpdateService(formState);
    }
  };
  // lifecycle hooks //
  useEffect(() => {
    if (serviceFormRef.current) {
      const elem = serviceFormRef.current.getBoundingClientRect();
      window.scrollTo({
        top: elem.bottom,
        behavior: "smooth"
      })
    }
  }, [serviceFormRef]);
  
  return (
    <div className="serviceFormDiv" ref={serviceFormRef}>
      <Form id="serviceForm">
        <Form.Field>
          <label>Service name</label>
          <input 
            id="adminServiceFormNameInput"
            onChange={handleNameChange} 
            placeholder="Service name here ..." 
            value={formState.name}
          />
        </Form.Field>
        <Form.Field>
          <label>Service price</label>
          <input 
            id="adminServiceFormPriceInput"
            onChange={handlePriceChange} 
            placeholder="Service price here..." 
            value={formState.price}
          />
        </Form.Field>
        <Form.Field
          id='adminServiceFormDescInput'
          control={TextArea}
          label='Store Description'
          onChange={handleDescriptionChange}
          placeholder='Service description here...'
          value={formState.description}
         />
         {
           newForm 
            ? <Button id="adminServiceFormCreate" type='submit' onClick={handleSubmit} content= "Create  New Service" />
            : <Button id="adminServiceFormUpdate" type='submit' onClick={handleSubmit} content= "Update Service" />
         }  
      </Form>
    </div>
  );
};

export default ServiceForm;