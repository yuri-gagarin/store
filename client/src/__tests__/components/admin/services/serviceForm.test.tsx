import React from "react";
import ServiceForm from "../../../../components/admin_components/services/forms/ServiceForm";
import { mount, ReactWrapper } from "enzyme";
// helpers //

describe("Service Form Component render tests", () => {
  describe("Empty Form render tests", () => {
    let wrapper: ReactWrapper;

    beforeAll(() => {
      window.scrollTo = jest.fn;
      // mock data //
      const serviceData: IServiceData = {
        _id: "",
        name: "",
        price: "",
        description: "",
        images: [],
        createdAt: ""
      }
      // mount form //
      wrapper = mount(
        <ServiceForm  
          name={serviceData.name}
          price={serviceData.price}
          description={serviceData.description}
          newForm={true}
          handleCreateService={jest.fn}
          handleUpdateService={jest.fn}
        />
      );
    });

    it("Should properly mount", () => {
      expect(wrapper).toMatchSnapshot();
    });
    it("Should render '#adminServiceFormNameInput'", () => {
      expect(wrapper.render().find("#adminServiceFormNameInput").length).toEqual(1);
    });
    it("Should render '#serviceFormPriceInput'", () => {
      expect(wrapper.render().find("#adminServiceFormPriceInput").length).toEqual(1);
    });
    it("Should render '#serviceFormDescInput'", () => {
      expect(wrapper.render().find("#adminServiceFormDescInput").length).toEqual(1);
    });
    it("Should render '#adminServiceFormCreate' Button", () => {
      expect(wrapper.render().find("#adminServiceFormCreateBtn").length).toEqual(1);
    });
    it("Should NOT render '#adminServiceFormUpdate' Button'", () => {
      expect(wrapper.render().find("#adminServiceFormUpdateBtn").length).toEqual(0);
    });
  });
  
  describe("Form with valid data render tests", () => {
    let wrapper: ReactWrapper; let mockData: IServiceData;

    beforeAll(() => {
      window.scrollTo = jest.fn;
      // mock service data //
      mockData = {
        _id: "1",
        name: "name",
        price: "100",
        description: "description",
        images: [],
        createdAt: "now"
      };

      wrapper = mount(
        <ServiceForm 
          name={mockData.name}
          price={mockData.price}
          description={mockData.description}
          newForm={false}
          handleCreateService={jest.fn}
          handleUpdateService={jest.fn}
        />
      );
    });

    it("Should properly mount", () => {
      expect(wrapper).toMatchSnapshot();
    });
    it("Should render '#adminServiceFormNameInput'", () => {
      expect(wrapper.render().find("#adminServiceFormNameInput").length).toEqual(1);
    });
    it("Should set correct data in '#adminServiceFormNameInput'", () => {
      const input = wrapper.render().find("#adminServiceFormNameInput");
      expect(input.val()).toEqual(mockData.name);
    });
    it("Should render '#adminServiceFormPriceInput'", () => {
      expect(wrapper.render().find("#adminServiceFormPriceInput").length).toEqual(1);
    });
    it("Should set correct data in '#adminServiceFormPriceInput'", () => {
      const input = wrapper.render().find("#adminServiceFormPriceInput");
      expect(input.val()).toEqual(mockData.price);
    });
    it("Should render '#adminServiceFormDescInput'", () => {
      expect(wrapper.render().find("#adminServiceFormDescInput").length).toEqual(1);
    });
    it("Should set correct data in '#adminServiceFormDescInput'", () => {
      const input = wrapper.render().find("#adminServiceFormDescInput");
      expect(input.val()).toEqual(mockData.description);
    });
    it("Should render '#adminServiceFormUpdate' Button'", () => {
      expect(wrapper.render().find("#adminServiceFormUpdateBtn").length).toEqual(1);
    });
    it("Should NOT render '#adminServiceFormCreate' Button", () => {
      expect(wrapper.render().find("#adminServiceFormCreateBtn").length).toEqual(0);
    });
  });
});