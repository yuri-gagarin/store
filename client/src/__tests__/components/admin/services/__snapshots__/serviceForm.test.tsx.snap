// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service Form Component render tests Empty Form render tests Should properly mount 1`] = `
<ServiceForm
  description=""
  handleCreateService={[Function]}
  handleUpdateService={[Function]}
  name=""
  newForm={true}
  price=""
>
  <div
    className="serviceFormDiv"
  >
    <Form
      as="form"
      id="serviceForm"
    >
      <form
        className="ui form"
        id="serviceForm"
        onSubmit={[Function]}
      >
        <FormField>
          <div
            className="field"
          >
            <label>
              Service name
            </label>
            <input
              id="adminServiceFormNameInput"
              onChange={[Function]}
              placeholder="Service name here ..."
              value=""
            />
          </div>
        </FormField>
        <FormField>
          <div
            className="field"
          >
            <label>
              Service price
            </label>
            <input
              id="adminServiceFormPriceInput"
              onChange={[Function]}
              placeholder="Service price here..."
              value=""
            />
          </div>
        </FormField>
        <FormField
          control={[Function]}
          id="adminServiceFormDescInput"
          label="Store Description"
          onChange={[Function]}
          placeholder="Service description here..."
          value=""
        >
          <div
            className="field"
          >
            <label
              htmlFor="adminServiceFormDescInput"
            >
              Store Description
            </label>
            <TextArea
              aria-describedby={null}
              as="textarea"
              id="adminServiceFormDescInput"
              onChange={[Function]}
              placeholder="Service description here..."
              rows={3}
              value=""
            >
              <Ref
                innerRef={
                  Object {
                    "current": <textarea
                      id="adminServiceFormDescInput"
                      placeholder="Service description here..."
                      rows="3"
                    />,
                  }
                }
              >
                <RefFindNode
                  innerRef={
                    Object {
                      "current": <textarea
                        id="adminServiceFormDescInput"
                        placeholder="Service description here..."
                        rows="3"
                      />,
                    }
                  }
                >
                  <textarea
                    aria-describedby={null}
                    id="adminServiceFormDescInput"
                    onChange={[Function]}
                    onInput={[Function]}
                    placeholder="Service description here..."
                    rows={3}
                    value=""
                  />
                </RefFindNode>
              </Ref>
            </TextArea>
          </div>
        </FormField>
        <Button
          as="button"
          content="Create  New Service"
          id="adminServiceFormCreate"
          onClick={[Function]}
          type="submit"
        >
          <Ref
            innerRef={
              Object {
                "current": <button
                  class="ui button"
                  id="adminServiceFormCreate"
                  type="submit"
                >
                  Create  New Service
                </button>,
              }
            }
          >
            <RefFindNode
              innerRef={
                Object {
                  "current": <button
                    class="ui button"
                    id="adminServiceFormCreate"
                    type="submit"
                  >
                    Create  New Service
                  </button>,
                }
              }
            >
              <button
                className="ui button"
                id="adminServiceFormCreate"
                onClick={[Function]}
                type="submit"
              >
                Create  New Service
              </button>
            </RefFindNode>
          </Ref>
        </Button>
      </form>
    </Form>
  </div>
</ServiceForm>
`;

exports[`Service Form Component render tests Form with valid data render tests Should properly mount 1`] = `
<ServiceForm
  description="description"
  handleCreateService={[Function]}
  handleUpdateService={[Function]}
  name="name"
  newForm={false}
  price="100"
>
  <div
    className="serviceFormDiv"
  >
    <Form
      as="form"
      id="serviceForm"
    >
      <form
        className="ui form"
        id="serviceForm"
        onSubmit={[Function]}
      >
        <FormField>
          <div
            className="field"
          >
            <label>
              Service name
            </label>
            <input
              id="adminServiceFormNameInput"
              onChange={[Function]}
              placeholder="Service name here ..."
              value="name"
            />
          </div>
        </FormField>
        <FormField>
          <div
            className="field"
          >
            <label>
              Service price
            </label>
            <input
              id="adminServiceFormPriceInput"
              onChange={[Function]}
              placeholder="Service price here..."
              value="100"
            />
          </div>
        </FormField>
        <FormField
          control={[Function]}
          id="adminServiceFormDescInput"
          label="Store Description"
          onChange={[Function]}
          placeholder="Service description here..."
          value="description"
        >
          <div
            className="field"
          >
            <label
              htmlFor="adminServiceFormDescInput"
            >
              Store Description
            </label>
            <TextArea
              aria-describedby={null}
              as="textarea"
              id="adminServiceFormDescInput"
              onChange={[Function]}
              placeholder="Service description here..."
              rows={3}
              value="description"
            >
              <Ref
                innerRef={
                  Object {
                    "current": <textarea
                      id="adminServiceFormDescInput"
                      placeholder="Service description here..."
                      rows="3"
                    >
                      description
                    </textarea>,
                  }
                }
              >
                <RefFindNode
                  innerRef={
                    Object {
                      "current": <textarea
                        id="adminServiceFormDescInput"
                        placeholder="Service description here..."
                        rows="3"
                      >
                        description
                      </textarea>,
                    }
                  }
                >
                  <textarea
                    aria-describedby={null}
                    id="adminServiceFormDescInput"
                    onChange={[Function]}
                    onInput={[Function]}
                    placeholder="Service description here..."
                    rows={3}
                    value="description"
                  />
                </RefFindNode>
              </Ref>
            </TextArea>
          </div>
        </FormField>
        <Button
          as="button"
          content="Update Service"
          id="adminServiceFormUpdate"
          onClick={[Function]}
          type="submit"
        >
          <Ref
            innerRef={
              Object {
                "current": <button
                  class="ui button"
                  id="adminServiceFormUpdate"
                  type="submit"
                >
                  Update Service
                </button>,
              }
            }
          >
            <RefFindNode
              innerRef={
                Object {
                  "current": <button
                    class="ui button"
                    id="adminServiceFormUpdate"
                    type="submit"
                  >
                    Update Service
                  </button>,
                }
              }
            >
              <button
                className="ui button"
                id="adminServiceFormUpdate"
                onClick={[Function]}
                type="submit"
              >
                Update Service
              </button>
            </RefFindNode>
          </Ref>
        </Button>
      </form>
    </Form>
  </div>
</ServiceForm>
`;
